### ISSUE TITLE
Remove unnecessary optional keywords and add documentation to `UpdateGridpoolOrderRequest`

### ISSUE THREAD

**Issue by thomas-nicolai-frequenz** [frequenz-floss/frequenz-api-electricity-trading#19]
### What's needed?

There are redundant optional keywords being used in the `UpdateGridpoolOrderRequest` message.

### Proposed solution

```Diff
// Request to update an existing order for a given Gridpool.
message UpdateGridpoolOrderRequest {
  // Represents the order properties that can be updated after an order has been
  // placed. At least one of the optional fields must be set for an update to
  // take place.
  message UpdateOrder {
-   // The updated limit price at which the contract is to be traded.
+   // Optional; The updated limit price at which the contract is to be traded.
    // This is the maximum price for a BUY order or the minimum price for a SELL
    // order.
-    optional frequenz.api.common.v1.market.Price price = 2;
+   frequenz.api.common.v1.market.Price price = 2;

-   // The updated quantity of the contract being traded, specified in MWh.
+  // Optional; The updated quantity of the contract being traded, specified in MWh.
-   optional frequenz.api.common.v1.market.Energy quantity = 3;
+  frequenz.api.common.v1.market.Energy quantity = 3;

-   // Applicable for STOP_LIMIT orders. This is the updated stop price that
+   // Optional; Applicable for STOP_LIMIT orders. This is the updated stop price that
    // triggers the limit order.
-   optional frequenz.api.common.v1.market.Price stop_price = 4;
+  frequenz.api.common.v1.market.Price stop_price = 4;

-   // Applicable for ICEBERG orders. This is the updated price difference
+   // Optional; Applicable for ICEBERG orders. This is the updated price difference
    // between the peak price and the limit price.
-   optional frequenz.api.common.v1.market.Price peak_price_delta = 5;
+  frequenz.api.common.v1.market.Price peak_price_delta = 5;

-   // Applicable for ICEBERG orders. This is the updated quantity of the order
+   // Optional; Applicable for ICEBERG orders. This is the updated quantity of the order
    // to be displayed in the order book.
-   optional frequenz.api.common.v1.market.Energy display_quantity = 6;
+  frequenz.api.common.v1.market.Energy display_quantity = 6;

-   // Updated execution options such as All or None, Fill or Kill, etc.
+   // Optional; Updated execution options such as All or None, Fill or Kill, etc.
    optional OrderExecutionOption execution_option = 7;

-   // This is an updated timestamp defining the time after which the order
+   // Optional; This is an updated timestamp defining the time after which the order
    // should be cancelled if not filled. The timestamp is in UTC.
    google.protobuf.Timestamp valid_until = 8;

-    // Updated user-defined payload individual to a specific order.
+   // Optional; Updated user-defined payload individual to a specific order.
    // This can be any data that the user wants to associate with the order.
    google.protobuf.Struct payload = 9;

-   // Updated user-defined tag to group related orders.
+   // Optional; Updated user-defined tag to group related orders.
    optional string tag = 10;
  }

  // ID of the Gridpool the order belongs to.
  uint64 gridpool_id = 1;

  // The order identifier
  uint64 order_id = 2;

  // Field mask specifying which fields should be updated
  google.protobuf.FieldMask update_mask = 3;

  // The fields that can be updated
  UpdateOrder update_order_fields = 4;
}

// ... 

// Represents an order in the electricity market.
message Order {
  // The delivery area where the contract is to be delivered. The representation
  // of the delivery area may vary by jurisdiction.
  frequenz.api.common.v1.grid.DeliveryArea delivery_area = 2;

  // The delivery period for the contract, specified as a start and end
  // timestamp in UTC. It represents the period during which the contract is
  // expected to be fulfilled.
  frequenz.api.common.v1.grid.DeliveryPeriod delivery_period = 3;

  // The type of order, such as LIMIT, STOP_LIMIT, ICEBERG etc.
  // This determines how the order is to be executed in the market.
  OrderType type = 4;

  // Indicates if the order is on the Buy or Sell side of the market.
  MarketSide side = 5;

  // The limit price at which the contract is to be traded. This is the maximum
  // price for a BUY order or the minimum price for a SELL order.
  frequenz.api.common.v1.market.Price price = 6;

  // The quantity of the contract being traded, specified in MWh.
  frequenz.api.common.v1.market.Energy quantity = 7;

- // Applicable for STOP_LIMIT orders. This is the stop price that triggers the
+ // Optional; Applicable for STOP_LIMIT orders. This is the stop price that triggers the
  // limit order.
-  optional frequenz.api.common.v1.market.Price stop_price = 8;
+ frequenz.api.common.v1.market.Price stop_price = 8;

- // Applicable for ICEBERG orders. This is the price difference between the
+ // Optional; Applicable for ICEBERG orders. This is the price difference between the
  // peak price and the limit price.
- optional frequenz.api.common.v1.market.Price peak_price_delta = 9;
+ frequenz.api.common.v1.market.Price peak_price_delta = 9;

-  // Applicable for ICEBERG orders. This is the quantity of the order to be
+  // Optional; Applicable for ICEBERG orders. This is the quantity of the order to be
  // displayed in the order book.
- optional frequenz.api.common.v1.market.Energy display_quantity = 10;
+ frequenz.api.common.v1.market.Energy display_quantity = 10;

  // Optional; Execution options such as All or None, Fill or Kill, etc.
  optional OrderExecutionOption execution_option = 11;

- // Do not use if ExecutionOption is set to FOK or IOC. This is an optional UTC
+ // Optional; Do not use if ExecutionOption is set to FOK or IOC. This is an optional UTC
  // timestamp defining the time after which the order should be cancelled if
  // not filled.
- optional google.protobuf.Timestamp valid_until = 12;
+ google.protobuf.Timestamp valid_until = 12;

  // Optional user-defined payload individual to a specific order. This can be
  // any data that needs to be associated with the order.
  //
  // The field can store e.g. JSON objects containing details involved in the
  // order. This feature can simplify application development by eliminating the
  // need for complicated state management to remember the specifics of each
  // order.
  //
  // By embedding this "state" within the order itself, you can include
  // specifics like which microgrids consume or provide how much power. This
  // makes it easier to manage complex orders and can simplify the logic
  // required in applications.
  //
  // Example JSON payload:
  // {
  //   "microgrids": [
  //     {
  //       "microgrid_id": "1",
  //       "mwh": 1.0
  //     },
  //     {
  //       "microgrid_id": "2",
  //       "mwh": 0.5
  //     }
  //   ]
  // }
  //
  // In this example, if the order is exectuted, these microgrids might consume
  // the electricity mentioned in the example JSON payload.
- optional google.protobuf.Struct payload = 13;
+ google.protobuf.Struct payload = 13;

  // Optional user-defined tag to group related orders.
  optional string tag = 14;
}

```

**Comment by tiyash-basu-frequenz** [comment]
The recommendation makes sense.

I'd also recommend adding that these fields are optional to the field documentations, so that users get informed of it when they use these fields in their IDEs/code-editors.

**Comment by thomas-nicolai-frequenz** [comment]
> fields are optional to the field documentations

Haven't I done this? Whats on your mind beyond of what I added to the documentation?

**Comment by tiyash-basu-frequenz** [comment]
>Haven't I done this? Whats on your mind beyond of what I added to the documentation?

Oh right, you did that already. I'd maybe extend it to add a suggestion what happens when it is left empty, wherever it is possible.

**Comment by ghost** [comment]
Why is the optional keyword  not removed for the stop price?

`optional frequenz.api.common.v1.market.Price stop_price = 8;`


And what is the 'sane default' of delivery_period and delivery_area? 

```
// Optional filter for delivery period.
optional frequenz.api.common.v1.grid.DeliveryPeriod delivery_period = 3;

// Optional filter for delivery area.
 optional frequenz.api.common.v1.grid.DeliveryArea delivery_area = 4;
```

_Update: These should not have the optional keyword (as their sane defaults are None). Will be updated._

**Comment by thomas-nicolai-frequenz** [comment]
> Why is the optional keyword not removed for the stop price?

That was a mistake. Its fixed now.

**Comment by thomas-nicolai-frequenz** [comment]
> And what is the 'sane default' of delivery_period and delivery_area?

There is none. If they are not specified they will be rejected. They have to be explicitly given. They are also not optional hence I'm not sure where you copied the lines above from. So far these are not optional.

It seems this does not belong to this issue as they are about filtering and not about creating or updating things.

**Comment by ghost** [comment]
> I'm not sure where you copied the lines above from


From the `GridpoolOrderFilter` [here](https://github.com/frequenz-floss/frequenz-api-electricity-trading/blob/v0.x.x/proto/frequenz/api/electricity_trading/v1/electricity_trading.proto#L399), and same applies to `PublicTradeFilter`. 
But yes I recon they should not have the optional keyword, I'll change it with the rest.

**Comment by thomas-nicolai-frequenz** [comment]
> But yes I recon they should not have the optional keyword, I'll change it with the rest.

No this is part of the filter. Both DeliveryPeriod and DeliveryArea are simple messages from my understanding and hence they can have the optional keyword in front of them. To me that looks correct. It means they do not need to be specified and if they are not specified there is no default! These are however used in combination with the field masks.

The default for an empty message is an empty message. That means they don't have a "default" as with an enum. The optional keyword enables field presence tracking. That means we can differentiate between a user explicitly given an empty message and saying I want this message to be reset. There is three states:

1. Setting the message but not including in the field mask
2. Setting a message and including in the field mask
3. Not setting a message but including it in the field mask

Each has different repurcushions but especially for enums because lets say we'd want to remove an enum value and NOT set it to its default. That would be impossible without the optional keyword. Now you are right for messages we might be able to remove it as we don't necessarily need the field presence tracking on the other handside it doesn't hurt knowing when something has been explicitly set and when not because an empty message might not equal an empty message.

More can be found [here](https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis).

**Comment by ghost** [comment]
> if they are not specified there is no default

I thought their default would precisely be None (since they are composite types). And thus the reason why the optional keyword was not needed here. 

> They can never be a "default" as with an enum

My understanding was that the 'sane default'  for enum is 0 or UNSPECIFIED as we have it. So that for these we needed to keep the optional keyword when needed.

**Comment by thomas-nicolai-frequenz** [comment]
> since they are composite types

Forget about the composite types stuff. Please have a look at this table [here](https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis).

**Comment by thomas-nicolai-frequenz** [comment]
You are right it can be removed. The optional keyword is redundant in this case as explicit field presence can be tracked wether the optional keyword is present or not for simple message types.

**Comment by matthias-wende-frequenz** [comment]
The changes have been merged in #21.
@camille-bouvy-frequenz can we close this issue?
